x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -232,6 +224,11 @@ private long readFileBytes(String fileName, BytesReference reference) throws IOE
                     }
                 });
 
+                long remainingSize = indexInput.length() - indexInput.getFilePointer();
+                if (remainingSize < reference.length()) {
+                    reference = reference.slice(0, Math.toIntExact(remainingSize));
```

Comment:
I'm not wild about the `read` method slicing internally and returning a different bytes reference. I also don't really understand why it is necessary as the follower should never request more data than is remaining in a file (you continue to only request the remaining bytes in this PR).

Reply:
The current API for getting a file is more iterator-oriented (i.e. read next, read next, read next), which is why I find it weird that the target will have to track information about where it is at in that file and make sure to request the right size. I have reverted this for now.